### PR TITLE
[region-isolation] When inferring isolation for an argument, handle non-self isolated parameters as well as self parameters that are actor isolated.

### PR DESCRIPTION
**Pull request by gottesmm**
[region-isolation] When inferring isolation for an argument, handle non-self isolated parameters as well as self parameters that are actor isolated.

As part of this I went through how we handled inference and rather than using a
grab-bag getActorIsolation that was confusing to use, I created split APIs for
specific use cases (actor instance, global actor, just an apply expr crossing)
that makes it clearer inside the SILIsolationInfo::get* APIs what we are
actually trying to model. I found a few issues as a result and fixed most of
them if they were small. I also fixed one bigger one around computed property
initializers in the next commit. There is a larger change I didn't fix around allowing function
ref/partial_apply with isolated self parameters have a delayed flow sensitive
actor isolation... this will be fixed in a subsequent commit.

This also fixes a bunch of cases where we were printing actor-isolated instead
of 'self' isolated.

rdar://127295657

----

NOTE: The third commit is the commit that fixes inferred isolation in property initializers in actors.

rdar://127844737

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: docs/SIL.rst [hunk 1]
```diff
@@ -692,6 +692,18 @@ autorelease in the callee.
   type behaves like a non-generic type, as if the substitutions were
   actually applied to the underlying function type.
 
+- SIL functions may optionally mark a function parameter as
+  ``@sil_isolated``. An ``@sil_isolated`` parameter must be one of:
+
+  - An actor or any actor type.
+  - A generic type that conforms to Actor or AnyActor.
```
Comment (ktoso) [inline docs/SIL.rst:699]: AnyActor being a marker protocol, I am wondering if this means Actor or DistributedActor in practice?

File: include/swift/SILOptimizer/Utils/PartitionUtils.h [hunk 2]
```diff
@@ -122,18 +181,29 @@ class SILIsolationInfo {
 
   /// If set this is the SILValue that represents the actor instance that we
   /// derived isolatedValue from.
-  SILValue actorInstance;
+  ///
+  /// If set to (SILValue(), 1), then we are in an
```
Comment (ktoso) [inline include/swift/SILOptimizer/Utils/PartitionUtils.h:185]: cut off sentence?

File: test/Concurrency/sendable_checking.swift [hunk 3]
```diff
@@ -97,7 +97,7 @@ public actor MyActor: MyProto {
   func g(ns1: NS1) async {
     await nonisolatedAsyncFunc1(ns1) // expected-targeted-and-complete-warning{{passing argument of non-sendable type 'NS1' outside of actor-isolated context may introduce data races}}
     // expected-tns-warning @-1 {{sending 'ns1' risks causing data races}}
-    // expected-tns-note @-2 {{sending actor-isolated 'ns1' to nonisolated global function 'nonisolatedAsyncFunc1' risks causing data races between nonisolated and actor-isolated uses}}
+    // expected-tns-note @-2 {{sending 'self'-isolated 'ns1' to nonisolated global function 'nonisolatedAsyncFunc1' risks causing data races between nonisolated and 'self'-isolated uses}}
```
Comment (ktoso) [inline test/Concurrency/sendable_checking.swift:100]: nice diagnostic improvement :)

File: lib/SIL/Verifier/SILVerifier.cpp [hunk 4]
```diff
@@ -6580,6 +6580,35 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+                                                     F.getTypeExpansionContext());
+        if (argType->isOptional())
+          argType = argType->lookThroughAllOptionalTypes()->getCanonicalType();
+
+        auto genericSig = FTy->getInvocationGenericSignature();
+        auto &ctx = F.getASTContext();
+        auto *actorProtocol = ctx.getProtocol(KnownProtocolKind::Actor);
+        auto *anyActorProtocol = ctx.getProtocol(KnownProtocolKind::AnyActor);
+        bool genericTypeWithActorRequirement = llvm::any_of(
+            genericSig.getRequirements(), [&](const Requirement &other) {
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SIL/Verifier/SILVerifier.cpp:6597]: Just talked with @xedin offline and he said it was reasonable.

File: docs/SIL.rst (hunk 1 above)
Comment (gottesmm) [reply docs/SIL.rst:699]: Not sure. What is important is that both have the semantics of Actors (namely they can create isolation).

**Pull request merged**